src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataDictionaryFormatter.cs

```diff
@@ -0,0 +1,363 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Composition.Formatter
+{
+    using System.Collections.Immutable;
+    using System.Globalization;
+    using System.Reflection;
+    using MessagePack;
+    using MessagePack.Formatters;
+    using Microsoft.VisualStudio.Composition.Reflection;
+
+#pragma warning disable CS8604 // Possible null reference argument.
+
+    internal class MetadataDictionaryFormatter : IMessagePackFormatter<IReadOnlyDictionary<string, object?>>
+    {
+        public static readonly MetadataDictionaryFormatter Instance = new();
+
+        private MetadataDictionaryFormatter()
+        {
+        }
+
+        /// <inheritdoc/>
+        public void Serialize(ref MessagePackWriter writer, IReadOnlyDictionary<string, object?> value, MessagePackSerializerOptions options)
+        {
+            writer.Write(value.Count);
```

Comment:
This is your one allowed written struct for the formatter.
What I think you meant was `writer.WriteMapHeader(value.Count);`, which would then give you license to write `value.Count * 2` more values.